crates/ruff/src/importer/insertion.rs

```diff
@@ -1,24 +1,90 @@
+#![allow(dead_code)]
+
 use ruff_text_size::TextSize;
 use rustpython_parser::ast::{Ranged, Stmt};
 use rustpython_parser::{lexer, Mode, Tok};
 
 use ruff_diagnostics::Edit;
 use ruff_python_ast::helpers::is_docstring_stmt;
 use ruff_python_ast::source_code::{Locator, Stylist};
+use ruff_textwrap::indent;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) enum Placement<'a> {
+    /// The content will be inserted inline with the existing code (i.e., within semicolon-delimited
+    /// statements).
+    Inline,
+    /// The content will be inserted on its own line.
+    OwnLine,
+    /// The content will be inserted as an indented block.
+    Indented(&'a str),
+}
 
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub(super) struct Insertion {
+pub(super) struct Insertion<'a> {
     /// The content to add before the insertion.
-    prefix: &'static str,
+    prefix: &'a str,
     /// The location at which to insert.
     location: TextSize,
     /// The content to add after the insertion.
-    suffix: &'static str,
+    suffix: &'a str,
+    /// The line placement of insertion.
+    placement: Placement<'a>,
 }
 
-impl Insertion {
-    /// Create an [`Insertion`] to insert (e.g.) an import after the end of the given [`Stmt`],
-    /// along with a prefix and suffix to use for the insertion.
+impl<'a> Insertion<'a> {
+    /// Create an [`Insertion`] to insert (e.g.) an import statement at the start of a given
+    /// file, along with a prefix and suffix to use for the insertion.
+    ///
+    /// For example, given the following code:
+    ///
+    /// ```python
+    /// """Hello, world!"""
+    ///
+    /// import os
+    /// ```
+    ///
+    /// The insertion returned will begin at the start of the `import os` statement, and will
+    /// include a trailing newline.
+    pub(super) fn start_of_file(
+        body: &[Stmt],
+        locator: &Locator,
+        stylist: &Stylist,
+    ) -> Insertion<'static> {
+        // Skip over any docstrings.
+        let mut location = if let Some(location) = match_docstring_end(body) {
+            // If the first token after the docstring is a semicolon, insert after the semicolon as an
+            // inline statement.
+            let first_token = lexer::lex_starts_at(locator.after(location), Mode::Module, location)
+                .flatten()
+                .next();
+            if let Some((Tok::Semi, range)) = first_token {
+                return Insertion::inline(" ", range.end(), ";");
+            }
+
+            // Otherwise, advance to the next row.
+            locator.full_line_end(location)
+        } else {
+            TextSize::default()
+        };
+
+        // Skip over any comments and empty lines.
+        for (tok, range) in
+            lexer::lex_starts_at(locator.after(location), Mode::Module, location).flatten()
+        {
+            if matches!(tok, Tok::Comment(..) | Tok::Newline) {
+                location = locator.full_line_end(range.end());
```

Comment:
Can we use `range.end()` directly when matching on `Tok::Newline | Tok::NonLogicalNewLines` and simply skipping comments?

Again, we may be able to implement this with a "trivial" text search.